src/ConfigPostProcessor.php

```diff
@@ -0,0 +1,107 @@
+<?php
+/**
+ * @see       https://github.com/laminas/laminas-zendframework-bridge for the canonical source repository
+ * @copyright https://github.com/laminas/laminas-zendframework-bridge/blob/master/COPYRIGHT.md
+ * @license   https://github.com/laminas/laminas-zendframework-bridge/blob/master/LICENSE.md New BSD License
+ */
+
+namespace Laminas\ZendFrameworkBridge;
+
+class ConfigPostProcessor
+{
+    /** @var Replacements */
+    private $replacements;
+
+    public function __construct()
+    {
+        $this->replacements = new Replacements([
+            // Since we're rewriting at the PHP level, and not the raw
+            // strings from reading file contents, we need to add this
+            // one.
+            'zend-expressive' => 'expressive',
+        ]);
+    }
+
+    /**
+     * @return array
+     */
+    public function __invoke(array $config)
+    {
+        $rewritten = [];
+
+        foreach ($config as $key => $value) {
+            $newKey = is_string($key) ? $this->replacements->replace($key) : $key;
+
+            if (isset($rewritten[$newKey]) && is_array($rewritten[$newKey])) {
```

Comment:
We are missing condition here that `$key !== $newKey`.

In case the key is exactly the same, and we are merging the contents we can duplicate values if arrays have numeric values. So I guess it will be bad for delegators:
```php
'delegators' => [
   Buzzer::class => [
        Delegator1Factory::class,
        Delegator2Factory::class,
   ],
],
```

the name `Buzzer::class` is gonna stay the same but in rewritten version we are gonna have:

```php
'delegators' => [
   Buzzer::class => [
        Delegator1Factory::class,
        Delegator2Factory::class,
        Delegator1Factory::class,
        Delegator2Factory::class,
   ],
],
```

unless - I am missing some condition somewhere?

Reply:
oh... and here will be another issue:
```php
'delegators' => [
   Buzzer::class => [
        \MyPackage\Delegator1::class,
        \Zend\Package\Delegator2::class,
   ],
],
```

it is gonna be rewritten and merged into:

```php
'delegators' => [
   Buzzer::class => [
        \MyPackage\Delegator1::class,
        \Zend\Package\Delegator2::class,
        \MyPackage\Delegator1::class,
        \Laminas\Package\Delegator2::class,
   ],
],
```

Reply:
The purpose of this clause is to detect if the key already exists in the replacements array, and is itself an array value; if it does, it indicates they have a keys named after both a ZF class AND its equivalent Laminas class. It likely should not happen, but this catches it, and, in that scenario, will _merge_ both arrays.

In terms of integer keys, you should never have top-level integer keys anyways, and, if you _do_ have integer keys, you cannot have duplicates.

However, that `isset()` should likely be an `array_key_exists` to ensure there is no type-munging. I can make that change. I'm not going to add a check for `$key !== $newKey` unless I have a test case to test against, though.

Reply:
> unless - I am missing some condition somewhere?

You are - this is happening _after_ configuration merging, so we won't have two keys of the same name, unless there are references to both a string AND its replacement. There will never be duplicate integer keys.

Reply:
Yes, you are correct.

So as discussed on Slack - 3rd party libraries can't really provide compatibility with ZF and Laminas/Expressive/Apigility packages. These must support ZF only (but because of our tools will be compatible with Laminas) or Laminas only, when our tools will be not applied.

Reply:
To summarize that discussion:

- 3rd party libraries get forwards compatibility for free when consumers use this post processor (as it will rewrite references to ZF classes)
- 3rd party libraries will eventually need to create a new major version that targets Laminas packages specifically, dropping support for the ZF versions.

Reply:
@weierophinney yes, sound good 👍